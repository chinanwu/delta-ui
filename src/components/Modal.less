@import '../style/colours';
@import '../style/shadows';

.Modal {
	position: fixed;
	top: 0;
	left: 0;
	right: 0;
	bottom: 0;
	height: 100%;
	width: 100%;
	display: flex;
	justify-content: center;
	align-items: center;
	background-color: @modalBg;
}

.Modal__content {
	background-color: white;
	color: black;
	box-shadow: @box-shadow-16;
	// Note: will need to deal with overflow by user of Modal. Since
	// modals usually should not be large enough for a overflow
}

.Modal__content--dark {
	background-color: @dark-10;
	color: #fff;
}

.Modal__close {
	position: absolute;
	top: 0;
	right: 0;
	font-size: 1.6rem;
	background-color: #d3d3d3;
	padding: 0.5rem 1rem;
	border: none;
	cursor: pointer;
	border-top-right-radius: 1rem;
	transition: background-color 0.2s;

	&:hover {
		background-color: #e4e4e4;
	}

	&:active {
		background-color: #b7b7b7;
	}
}

.Modal__close--dark {
	background-color: #4d4d4d;
	color: #eeeeee;

	&:hover {
		background-color: #636363;
	}

	&:active {
		background-color: #3a3a3a;
	}
}
