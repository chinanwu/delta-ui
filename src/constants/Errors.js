export const ERROR_NOT_ONE_OFF =
	'Word must be one letter off from previous word';
export const ERROR_NOT_REAL_WORD = "Word entered isn't a real word";
export const ERROR_INVALID_WORD_ENTERED = 'Invalid word entered';
export const ERROR_INVALID_WORDS_ENTERED = 'Invalid word(s) entered';
