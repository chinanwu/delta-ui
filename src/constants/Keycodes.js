export const backspaceBtn = 8;
export const tabBtn = 9;
export const enterBtn = 13;
export const spaceBtn = 32;
export const rightArrowBtn = 39;
export const leftArrowBtn = 37;
export const upArrow = 38;
export const downArrow = 40;
export const aBtn = 65;
export const zBtn = 90;
export const escapeBtn = 27;
