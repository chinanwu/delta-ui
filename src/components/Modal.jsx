import FocusTrap from 'focus-trap-react';
import PropTypes from 'prop-types';
import React, { useCallback } from 'react';

import { escapeBtn } from '../constants/Keycodes';
import getThemeClassname from '../functions/getThemeClassname';

import './Modal.less';

export const Modal = ({
	dark,
	name,
	ariaLabelledBy,
	contentClassname,
	general,
	children,
	onClose,
}) => {
	const handleKeyDown = useCallback(
		event => {
			if (onClose) {
				if (event && event.keyCode) {
					if (event.keyCode === escapeBtn) {
						event.preventDefault();
						onClose();
					}
				}
			}
		},
		[onClose]
	);

	const handleClose = useCallback(() => {
		onClose();
	}, [onClose]);

	return (
		<FocusTrap>
			<div className="Modal" onKeyDown={handleKeyDown}>
				{general}
				<div
					className={
						getThemeClassname('Modal__content', dark) +
						(contentClassname ? ' ' + contentClassname : '')
					}
					role="dialog"
					aria-modal={true}
					aria-labelledby={ariaLabelledBy}
				>
					{onClose && (
						<button
							id="modalClose"
							className={getThemeClassname('Modal__close', dark)}
							aria-label={`Close ${name} modal`}
							onClick={handleClose}
						>
							X
						</button>
					)}

					{children}
				</div>
			</div>
		</FocusTrap>
	);
};

Modal.propTypes = {
	dark: PropTypes.bool,
	name: PropTypes.string,
	contentClassname: PropTypes.string,
	ariaLabelledBy: PropTypes.string,
	general: PropTypes.any,
	children: PropTypes.any,
	onClose: PropTypes.func,
};

export default Modal;
