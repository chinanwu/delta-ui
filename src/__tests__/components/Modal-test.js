import { shallow } from 'enzyme';
import React from 'react';

import { Modal } from '../../components/Modal.jsx';
import { aBtn, escapeBtn } from '../../constants/Keycodes';

jest.unmock('../../components/Modal.jsx');
jest.unmock('../../functions/getThemeClassname');

describe('Modal component', () => {
	describe('rendering', () => {
		it('renders Modal', () => {
			const wrapper = shallow(<Modal />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders close button', () => {
			const wrapper = shallow(<Modal onClose={jest.fn()} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders additional styling', () => {
			const wrapper = shallow(<Modal contentClassname="Test" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('closes Modal on click', () => {
			const onClose = jest.fn();
			const wrapper = shallow(<Modal onClose={onClose} />);
			expect(wrapper.getElement()).toMatchSnapshot();
			wrapper.find('#modalClose').simulate('click');
			expect(onClose).toHaveBeenCalledTimes(1);
		});

		it('closes Modal with escape key press', () => {
			const preventDefault = jest.fn();
			const event = {
				keyCode: escapeBtn,
				preventDefault,
			};
			const onClose = jest.fn();
			const wrapper = shallow(<Modal id="test" onClose={onClose} />);
			expect(wrapper.getElement()).toMatchSnapshot();
			wrapper.find('#test').simulate('keydown', event);
			expect(onClose).toHaveBeenCalledTimes(1);
			expect(preventDefault).toHaveBeenCalledTimes(1);
		});

		it('does nothing other key press', () => {
			const event = {
				keyCode: aBtn,
			};
			const onClose = jest.fn();
			const wrapper = shallow(<Modal id="test" onClose={onClose} />);
			expect(wrapper.getElement()).toMatchSnapshot();
			wrapper.find('#test').simulate('keydown', event);
			expect(onClose).not.toHaveBeenCalled();
		});

		it('does nothing with invalid event', () => {
			const onClose = jest.fn();
			const wrapper = shallow(<Modal id="test" onClose={onClose} />);
			expect(wrapper.getElement()).toMatchSnapshot();
			wrapper.find('#test').simulate('keydown');
			expect(onClose).not.toHaveBeenCalled();
		});

		it('does nothing if key press but no close', () => {
			const wrapper = shallow(<Modal id="test" />);
			expect(wrapper.getElement()).toMatchSnapshot();
			wrapper.find('#test').simulate('keydown');
		});
	});
});
