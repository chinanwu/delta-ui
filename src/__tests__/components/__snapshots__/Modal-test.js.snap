// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component behaviour closes Modal 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    >
      <button
        aria-label="Close undefined modal"
        className="Modal__close"
        id="modalClose"
        onClick={[Function]}
      >
        X
      </button>
    </div>
  </div>
</FocusTrap>
`;

exports[`Modal component behaviour closes Modal on click 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    >
      <button
        aria-label="Close undefined modal"
        className="Modal__close"
        id="modalClose"
        onClick={[Function]}
      >
        X
      </button>
    </div>
  </div>
</FocusTrap>
`;

exports[`Modal component behaviour closes Modal with escape key press 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    id="test"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    >
      <button
        aria-label="Close undefined modal"
        className="Modal__close"
        id="modalClose"
        onClick={[Function]}
      >
        X
      </button>
    </div>
  </div>
</FocusTrap>
`;

exports[`Modal component behaviour does nothing if key press but no close 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    id="test"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    />
  </div>
</FocusTrap>
`;

exports[`Modal component behaviour does nothing other key press 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    id="test"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    >
      <button
        aria-label="Close undefined modal"
        className="Modal__close"
        id="modalClose"
        onClick={[Function]}
      >
        X
      </button>
    </div>
  </div>
</FocusTrap>
`;

exports[`Modal component behaviour does nothing with invalid event 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    id="test"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    >
      <button
        aria-label="Close undefined modal"
        className="Modal__close"
        id="modalClose"
        onClick={[Function]}
      >
        X
      </button>
    </div>
  </div>
</FocusTrap>
`;

exports[`Modal component rendering renders Loading 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    />
  </div>
</FocusTrap>
`;

exports[`Modal component rendering renders Modal 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    />
  </div>
</FocusTrap>
`;

exports[`Modal component rendering renders additional styling 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content Test"
      role="dialog"
    />
  </div>
</FocusTrap>
`;

exports[`Modal component rendering renders close button 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={true}
  focusTrapOptions={Object {}}
  paused={false}
>
  <div
    className="Modal"
    onKeyDown={[Function]}
  >
    <div
      aria-modal={true}
      className="Modal__content"
      role="dialog"
    >
      <button
        aria-label="Close undefined modal"
        className="Modal__close"
        id="modalClose"
        onClick={[Function]}
      >
        X
      </button>
    </div>
  </div>
</FocusTrap>
`;
